utils/utils.py

```diff
@@ -0,0 +1,299 @@
+    wordlist_B = sentence_B.split(" ")
+    for word in wordlist_A:
+      if word not in wordlist:
+        strword = '['+ word + ']'
+        wordlist.append(strword)
+    for word in wordlist_B:
+      if word not in wordlist:
+        strword = '[' + word + ']'
+        wordlist.append(strword)
+    sigma = pynini.accep("")
+    for word in wordlist:
+      sigma = sigma.union(pynini.accep(word))
+    sigma = sigma.optimize()
+
+    insert = pynutil.insert(f"[{self.INSERT}]", weight=insert_cost / 2)
+    delete = pynini.cross(
+        sigma, pynini.accep(f"[{self.DELETE}]", weight=delete_cost / 2)) 
+    substitute = pynini.cross(
+        sigma, pynini.accep(f"[{self.SUBSTITUTE}]", weight=substitute_cost / 2)) 
+
+    edit = pynini.union(insert, delete, substitute).optimize()
+    if bound:
+      sigma_star = pynini.closure(sigma)
+      self._e_i = sigma_star.copy()
+      for _ in range(bound):
+        self._e_i.concat(edit.ques).concat(sigma_star)
+    else:
+      self._e_i = edit.union(sigma).closure()
+  
+    self._e_i.optimize()
+    self._e_o = EditTransducer._right_factor(self._e_i)
+
+  @staticmethod
+  def _right_factor(ifst: pynini.Fst) -> pynini.Fst:
+    """Constructs the right factor from the left factor."""
+    # Ts constructed by inverting the left factor (i.e., swapping the input and
+    # output labels), then swapping the insert and delete labels on what is now
+    # the input side.
+    ofst = pynini.invert(ifst)
+    syms = pynini.generated_symbols()
+    insert_label = syms.find(EditTransducer.INSERT)
+    delete_label = syms.find(EditTransducer.DELETE)
+    pairs = [(insert_label, delete_label), (delete_label, insert_label)]
+    return ofst.relabel_pairs(ipairs=pairs)
+
+  @staticmethod
+  def check_wellformed_lattice(lattice: pynini.Fst) -> None:
+    """Raises an error if the lattice is empty.
+    Args:
+      lattice: A lattice FST.
+    Raises:
+      Error: Lattice is empty.
+    """
+    if lattice.start() == pynini.NO_STATE_ID:
+      raise Error("Lattice is empty")
+
+  def create_lattice(self, iexpr: pynini.FstLike,
+                     oexpr: pynini.FstLike) -> pynini.Fst:
+    """Creates edit lattice for a pair of input/output strings or acceptors.
+    Args:
+      iexpr: input string or acceptor
+      oexpr: output string or acceptor.
+    Returns:
+      A lattice FST.
+    """
+    sybt = dict()
+    syn_list = load_labels(get_abs_path("synonym.tsv"))
+    for wordlist in syn_list:
+      for word in wordlist:
+        if word in iexpr:# syn found
+          if len(word.split(" "))>1:#syn with more than one word's
+            wordstring = word.split(" ")[0]
+            for j in range(1,len(word.split(" "))):
+              wordstring+="_"+word.split(" ")[j]
+            tagger = pynini.cross(word,wordstring)
+            tagger = pynini.cdrewrite(tagger,"","",pynini.closure(utf8.VALID_UTF8_CHAR))        
+            iexpr = (iexpr @ tagger).string()
+    iex_list = iexpr.split(" ")
```

Comment:
code-reorg: these code (converting ref & hyp from text form to fst form) should be moved outside, and create_lattice is responsible only for:

1. accepting well-structured iexpr & oexpr fsts
2. apply core composition algorithm to expand the WER alignment space(i.e. lattice). 

This also raises an important question: I have a feeling that we don't need to touch lower level algorithm code, say things like this `lattice = (iex @ self._e_i) @ (self._e_o @ oex)` should not appear in our code base.  Can we just import pynin's library and make calls to provided interfaces? that will greatly simplify our code.